Refactor bolt server to easier allow multiple protocol versions

lutovich opened the pull request:
PR makes it easier to introduce new bolt protocol versions. It adds a factory for protocol handlers and extracts interface for V1 format so that it can be easier extended with new types.

lutovich committed:
Extracted factory for bolt protocol handlers

Previously mapping from protocol version to a function that created
protocol handler was kept in a map. This map was also passed around
as-is. It was not pretty because map type was rather large and it
was mutable at any point.

This commit extracts a dedicated factory that is able to create
protocol handlers for the given version and network channel. Code
looks cleaner with this factory.

lutovich committed:
Extracted interface for Neo4jPack

To make it easier to plug different version that, for example, supports
new types, like point, geometry or date.

lutovich committed:
Small cleanup around Neo4jPack

Removed unneeded and unused methods and unified exception propagation
between Packer and Unpacker.

pontusmelke commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/Neo4jPackV1.java:
```diff
@@ -0,0 +1,609 @@
+        @Override
+        public void writeNull() throws IOException
+        {
+            packNull();
```
Maybe we could inline all these method calls, not sure if there is a benefit of having both `pack( X value )` and `writeX( X value )`?

lutovich replied:
You are right, having both `#pack()` and `#write()` methods is confusing. Former come from `PackStream.Packer` which this class extends. So this class is a mix of high level methods like `#pack(AnyValue)` and low level packstream methods like `#packStructHeader()`.

I think simple inlining will make things even more confusing. This class will have to deal with `PackOutput` directly and take on some packstream functionality.

zhenlineo replied:
@pontusmelke Is it correct that`Neo4jPackV1.Packer` is actually a bolt impl of `AnyValueWriter` and `PackStream.Packer` is a bolt impl of `ValueWriter`?

Then maybe the long term cleanup plan would be make PackStream.Packer impl `ValueWriter` and `Neo4jPackV1.Packer` impl `AnyValueWriter`?

pontusmelke commented on community/bolt/src/test/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageWriter.java:
```diff
@@ -58,7 +58,7 @@ public void onInit( String clientName, Map<String,Object> credentials ) throws I
     {
         packer.packStructHeader( 1, INIT.signature() );
```
This is actually wrong but it is an error that has propagated to all drivers. The init struct has size 2 and not size 1. We are saved by ignoring the struct size when we read the init message.

lutovich replied:
True :)
We do not really use struct size for predefined messages like INIT or RUN. It is kind of ignored [here](https://github.com/neo4j/neo4j/blob/3.4.0-alpha04/community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/BoltRequestMessageReader.java#L58). Code decides how to unpack based on the struct signature. Worth fixing anyway, I'll create cards.

lutovich committed:
One more test for AuthTokenValuesWriterTest

To verify that same writer can be used to write
multiple different values.

lutovich committed:
Use array when unpacking list of known size

Bolt server can unpack lists of unknown size. It stores all unpacked
values in an `ArrayList` which automatically grows as elements are
added. Then list is converted to an array. Same thing was done for
lists of known size.

This commit removes intermediate `ArrayList` so that only a single
array is allocated.

lutovich committed:
Fixed INIT struct size

INIT struct has two fields - user agent and auth token. Thus size of
the INIT struct is 2. However it has been packed as 1 by the
`BoltRequestMessageWriter`, which is only used in tests.

This commit fixes the problem.

zhenlineo reviewed (commented):
Vote for keeping as it is and we will try to clean in later versions.
